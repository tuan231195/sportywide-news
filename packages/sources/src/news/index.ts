export * from './au-news';
export * from './cnet-news';
export * from './smh-news';
export * from './tech-republic-news';
export * from './factory';
export * from './utils';
