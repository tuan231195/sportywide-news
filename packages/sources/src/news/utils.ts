import axios, { AxiosInstance } from 'axios';
import Cheerio from 'cheerio';
import { JSDOM } from 'jsdom';
import createDOMPurify from 'dompurify';
import adapter from 'axios/lib/adapters/http';
import htmlToText from 'html-to-text';

const window = new JSDOM('').window;
const DOMPurify = createDOMPurify(window);

export function getParsedXml(axios: AxiosInstance, url) {
	return axios.get(url).then(({ data, headers }) => {
		const contentType = headers['content-type'];
		if (
			contentType &&
			[
				'application/xml',
				'application/rss+xml',
				'text/xml; charset=UTF-8',
				'text/xml',
			].includes(contentType)
		) {
			return Cheerio.load(data, {
				xmlMode: true,
			});
		}
		return data;
	});
}

export function resolveLazyLoadedImage($: CheerioStatic, source: Cheerio) {
	source.find('img').each((index, element) => {
		const dataSource = $(element).data('src');
		const currentSource = $(element).attr('src');
		if (dataSource && !currentSource) {
			$(element).attr('src', $(element).data('src'));
		}
	});
	return source;
}

export function getCleanedHTML(source) {
	return DOMPurify.sanitize(source, {
		ALLOW_DATA_ATTR: false,
		FORBID_ATTR: ['style', 'class', 'id', 'width', 'height'],
	});
}

export function getThumbnailUrl(source) {
	const $ = Cheerio.load(source);
	return $('img').eq(0).attr('src');
}

export function getRawText(source) {
	return htmlToText.fromString(source, {
		wordwrap: 130,
		format: {
			text: function (elem) {
				if (elem.parent?.name === 'div') {
					return `${elem.data}\n`;
				} else if (elem.parent?.name === 'span') {
					return `${elem.data} `;
				}
				return elem.data;
			},
		},
	});
}

axios.defaults.adapter = adapter;
export { axios };
