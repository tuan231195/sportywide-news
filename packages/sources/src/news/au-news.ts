import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import { axios, resolveLazyLoadedImage } from './utils';
import Cheerio from 'cheerio';
import { getCleanedHTML } from 'src/news/utils';
import { arr } from '@vdtn359/news-utils';

export class AuNews extends DefaultNews {
	static url = 'www.news.com.au';

	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	static async extractUrl(url: string): Promise<string> {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		const contentSelectors = ['.story-content', '.story-body'];
		let article = arr.findMap(
			contentSelectors,
			(selector) => $(selector),
			(element) => !!element?.length
		);
		article = resolveLazyLoadedImage($, article);
		const storyContent = Cheerio.html(article);
		return getCleanedHTML(storyContent);
	}
}
