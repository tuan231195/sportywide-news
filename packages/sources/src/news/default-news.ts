import { News } from 'src/news.interface';
import { CATEGORY, getSlug, NewsDto } from '@vdtn359/news-models';
import { getParsedXml } from 'src/news/utils';
import axios, { AxiosInstance } from 'axios';
import { str } from '@vdtn359/news-utils';
import { parse } from 'date-fns';

export abstract class DefaultNews implements News {
	protected axios: AxiosInstance;
	protected constructor(
		protected readonly category: CATEGORY,
		protected rssFeed: string
	) {
		this.axios = axios.create({
			baseURL: rssFeed,
		});
	}

	async extractFeeds(): Promise<NewsDto[]> {
		const $: CheerioStatic = await getParsedXml(this.axios, this.rssFeed);
		return Array.from($('item')).map((node) => {
			const element = $(node);
			const url = element.find('> link').text().trim();
			const title = element.find('> title').text().trim();
			const description = element.find('> description').text().trim();
			const pubDate =
				element.find('pubDate').text() || $('pubDate').text();

			return {
				category: this.category,
				id: str.toGuid(url),
				title,
				description,
				image: this.getImage(element),
				pubDate: parse(
					pubDate,
					'E, dd MMM yyyy HH:mm:ss XX',
					new Date()
				),
				url,
				slug: getSlug(url),
				feed: this.rssFeed.trim(),
			};
		});
	}

	getImage(element: Cheerio): string | undefined {
		const image = element.find('image');
		return image.length ? image.find('url').text().trim() : undefined;
	}
}
