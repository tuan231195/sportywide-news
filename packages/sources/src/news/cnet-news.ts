import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import Cheerio from 'cheerio';
import { getCleanedHTML, resolveLazyLoadedImage } from 'src/news/utils';
import { axios } from './utils';

export class CnetNews extends DefaultNews {
	static url = 'www.cnet.com';
	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	getImage(element: Cheerio): string | undefined {
		const image = element.find('media\\:thumbnail');
		return image.length ? image.attr('url')! : undefined;
	}

	static async extractUrl(url: string) {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		let storyBody = $('#article-body');
		storyBody.find('figure > a > svg').parent().remove();
		const toRemove = [
			'[data-video-playlist]',
			'[section="shortcodeGallery"]',
			'footer',
			'.playerControls',
			'[data-component="lazyloadElement"]',
		];
		toRemove.forEach((section) => {
			storyBody.find(section).remove();
		});
		storyBody = resolveLazyLoadedImage($, storyBody);
		const storyContent = Cheerio.html(storyBody);
		return getCleanedHTML(storyContent);
	}
}
