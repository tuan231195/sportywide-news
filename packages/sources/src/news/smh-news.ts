import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import { axios, getCleanedHTML, resolveLazyLoadedImage } from 'src/news/utils';
import Cheerio from 'cheerio';
import { arr } from '@vdtn359/news-utils';

export class SmhNews extends DefaultNews {
	static url = 'www.smh.com.au';

	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	static async extractUrl(url: string) {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		const contentSelectors = [
			'#content > div > div > article > div > div:nth-child(1) > div:nth-child(2)',
			'#content > div > article > section > div',
			'#content > div > article',
		];
		let articleBody = arr.findMap(
			contentSelectors,
			(selector) => $(selector),
			(element) => !!element?.length
		);
		if (!articleBody?.length) {
			return '';
		}
		const toRemove = ['#endOfArticle', '.noPrint', '[preload=metadata]'];
		toRemove.forEach((section) => {
			articleBody.find(section).remove();
		});
		articleBody = resolveLazyLoadedImage($, articleBody);
		const storyContent = Cheerio.html(articleBody);
		return getCleanedHTML(storyContent);
	}
}
