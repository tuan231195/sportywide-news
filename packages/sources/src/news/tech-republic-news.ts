import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import { axios, getCleanedHTML, resolveLazyLoadedImage } from 'src/news/utils';
import Cheerio from 'cheerio';

export class TechRepublicNews extends DefaultNews {
	static url = 'www.techrepublic.com';

	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	static async extractUrl(url: string) {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		let articleBody = $('article.article-single > div.content');
		const toRemove = ['[data-component="newsletterSubscription"]'];
		toRemove.forEach((section) => {
			articleBody.find(section).remove();
		});
		articleBody = resolveLazyLoadedImage($, articleBody);
		const storyContent = Cheerio.html(articleBody);
		return getCleanedHTML(storyContent);
	}
}
