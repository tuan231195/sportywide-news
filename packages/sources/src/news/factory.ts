import { CnetNews } from 'src/news/cnet-news';
import { AuNews } from 'src/news/au-news';
import { TechRepublicNews } from 'src/news/tech-republic-news';
import { SmhNews } from 'src/news/smh-news';
import { keyBy } from 'lodash';
import urlParser from 'url';

export const availableSources = [AuNews, CnetNews, TechRepublicNews, SmhNews];
const sourceMap = keyBy(availableSources, 'url');

export function extractUrl(url) {
	const parsedUrl = urlParser.parse(url);
	const host: any = parsedUrl.hostname;
	const source = sourceMap[host];
	if (!source) {
		return;
	}
	return source.extractUrl(url);
}
