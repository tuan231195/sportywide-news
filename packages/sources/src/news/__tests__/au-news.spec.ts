import { getRawText } from 'src/news/utils';
import { AuNews } from 'src/news';

describe.skip('#extractNews', () => {
	it('should strip style and css and data attributes', async () => {
		expect(
			await AuNews.extractUrl(
				'https://www.news.com.au/travel/travel-advice/coronavirus-in-australia-mandatory-quarantine-in-hotels-for-anyone-returning-from-overseas/news-story/353e68837a888122a223d53bb2e4cf03'
			)
		).toMatchSnapshot();
	});
});
