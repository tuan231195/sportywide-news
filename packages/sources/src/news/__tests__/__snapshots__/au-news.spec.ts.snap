// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#extractNews should strip style and css and data attributes 1`] = `"<div><p>Prime Minister Scott Morrison has announced strict new isolation measures for those who have come home from abroad, which will be in place by Saturday at midnight, in a bid to slow the spread of <a type=\\"[object Object]\\" href=\\"https://www.news.com.au/topics/coronavirus\\">coronavirus</a>.</p><p>“States and territories will be quarantining all arrivals through our airports in hotels and other accommodation facilities for the two weeks of their mandatory self-isolation before they are able to return to their home,” Mr Morrison said at a press conference in Canberra.</p><p>“If their home is in South Australia or Perth or Tasmania and they’ve arrived in Melbourne, they will be quarantining in Melbourne.”</p><p>Mr Morrison said the strict new measure was necessary given two-thirds of all confirmed cases of coronavirus in Australia are the result of Aussies coming home from overseas.</p><p>And to ensure people remain in the hotels where they’re placed for 14 days, the Commonwealth will provide support by deploying the Australian Defence Force “to assist with compliance”, he said.</p><p><b>RELATED: <a type=\\"[object Object]\\" href=\\"https://www.news.com.au/topics/coronavirus\\">Follow the latest coronavirus updates</a></b></p><div><div><img alt=\\"Australians abroad continue to make their way home, but from Saturday, they’ll be made to spend 14 days in quarantine.\\"></div><p>Australians abroad continue to make their way home, but from Saturday, they’ll be made to spend 14 days in quarantine.<span><em>Source:News Corp Australia</em></span></p></div><div><div><img alt=\\"Scott Morrison announces strict new isolation measure.\\"></div><p>Scott Morrison announces strict new isolation measure.<span><em>Source:SKY</em></span></p></div><p>The PM was at pains to stress that the military had not been authorised to prosecute people for breaches – that remains the responsibility of law enforcement agencies in each jurisdiction.</p><p>“The ADF will be there to support those enforcement authorities. And so, we will be turning to the Defence Forces to support compliance with these new arrangements.</p><div></div><p>“It will require cooperative and facilitative support, and I have no doubt the Forces will do that in the most sensitive way they can, but it is necessary.”</p><p>For some time now, anyone arriving in Australia from overseas has been required to fill out an isolation declaration card, detailing where they will be self-isolating.</p><p>“This is enforceable by law,” Mr Morrison said.</p><p>There have reports of breaches of those self-isolation requirements and states and territories have stepped up their enforcement protocols.</p><p>“If you have come back into Australia, you need to live up to this pledge and the state and territory governments are going to make sure you do.”</p><p>But people were previously able to do that at home – something that ends tomorrow.</p><p><b>RELATED: <a type=\\"[object Object]\\" href=\\"https://www.news.com.au/lifestyle/health/health-problems/coronavirus-simple-graph-shows-how-we-can-flatten-the-curve-and-lessen-the-disruption-caused-by-the-pandemic/news-story/6acd8c4c5c453e6138228b2ab2dde008\\">How flattening the curve saves lives</a></b></p><div><div><img alt=\\"Australians returning from abroad will be forced into quarantine for 14 days. Picture: AAP\\"></div><p>Australians returning from abroad will be forced into quarantine for 14 days. Picture: AAP<span><em>Source:AAP</em></span></p></div><p>“For those who are seeking to make their way back, know these arrangements will be in place and they are mandatory,” Mr Morrison said.</p><p>“They should also know that it won't be before too long where we will be very difficult to get back do Australia.”</p><p>In terms of the number of returning Australians likely to be quarantined, the PM said 7120 people arrived at airports across the country yesterday.</p><p>The same new arrangement will apply to the Vasco da Gama cruise ship that’s currently off Perth, with a number of Australians on board.</p><p>The West Australian Government plans to quarantine those passengers on Rottnest Island, which was this week cleared to make way for them next week.</p><div><div><img alt=\\"The Vasco de Gama cruise ship is currently off Perth.\\"></div><p>The Vasco de Gama cruise ship is currently off Perth.<span><em>Source:Getty Images</em></span></p></div><p>Speaking in Melbourne this afternoon, Victorian Premier Daniel Andrews said it didn’t matter where Australians were returning from given coronavirus is “running so wildly rampant in a number of other countries”.</p><p>“If you have been in international airports and are returning from anywhere, really, you are an unacceptable risk if we just let you go about your business,” Mr Andrews said.</p><p>“You have got to quarantine, and we are now going to have a situation from midnight tomorrow night where you will not be quarantining at home on an honesty basis.</p><p>“You will be quarantining with appropriate support at a hotel with everything, here in the CBD, in the suburbs or in regional areas, and we will make those arrangements as convenient as we can.</p><p>“It is a big step to take away someone's liberty and make them go to a certain place and stay there for two weeks but this is life-and-death.”</p><p>Between 1300 and 2000 people are arriving back into Melbourne each day and the Victorian Government has 5000 hotel rooms on standby.</p><p>“We will do the work to step that up,” Mr Andrews said.</p><p>“Those numbers (of arrival) will perhaps drop off in coming days but we have sufficient capacity and we have done the work early to be able to accommodate those people.</p><p>“They will be housed, fed and transported at no cost to them but this is the best way to ensure absolute compliance.”</p><p>Some 500 police officers are conducting self-isolation enforcement now for those who have arrived back in recent days and weeks to ensure people are where they should be. “So, if you are doing the wrong thing you will be caught.”</p><div><div><img alt=\\"Victorian Premier Daniel Andrews says the new quarantine measures as a matter of “life and death”.\\"></div><p>Victorian Premier Daniel Andrews says the new quarantine measures as a matter of “life and death”.<span><em>Source:News Corp Australia</em></span></p></div><p>Chief Medical Officer Professor Brendan Murphy said health authorities are “worried” about the continued rapid growth in infection numbers, which today tipped over 3000.</p><p>“But as the Prime Minister said we're in an almost unique situation in this country in that even now a substantial part of the new cases is returned travellers,” Professor Murphy said.</p><p>“Our numbers have not been helped by a recent cruise ship, which has given lots of cases in Sydney. We are still seeing significant numbers of returned travellers with the virus developing an infection and, in many cases, passing it on to their families.”</p><p>It’s critical to “stop the capacity” for any returning travellers to spread coronavirus, Professor Murphy.</p><p>“We would look after them when they get the virus, as some will do, as they continue to come home, coming from countries now that have large outbreaks, so the risk is getting higher and higher as the number of flights reduce, but the risk in those countries increase.</p><p>“ This is a really important thing. But we are also worried about community transmission. That is the single most important concern for the health experts.”<br></p></div>"`;
